File: 695_MaxAreaofIsland.java
class Solution {
    int area = 0;
    public int maxAreaOfIsland(int[][] grid) {
        if(grid == null || grid.length == 0) return 0;
        int n = grid.length;
        int m = grid[0].length;
        int res = 0;
        boolean[][] vis = new boolean[n][m];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (vis[i][j] || grid[i][j] == 0) {
                    continue;
                }
                area = 0;
                dfs(i, j, grid, vis, n, m);
                res = Math.max(area, res);
            }
        }

        return res;
    }
    void dfs(int i, int j, int[][] grid, boolean[][] vis, int n, int m) {
        if (i < 0 || i >= n || j < 0 || j >= m || vis[i][j]) {
            return;
        }
        if (grid[i][j] == 0) {
            vis[i][j] = true;
            return;
        }
        vis[i][j] = true;
        area++;
        dfs(i+1, j, grid, vis, n, m);
        dfs(i-1, j, grid, vis, n, m);
        dfs(i, j+1, grid, vis, n, m);
        dfs(i, j-1, grid, vis, n, m);
    }
}

File: c++/695_MaxAreaofIsland.cpp
class Solution {
public:
    int dx[4] = {1, 0, -1, 0};
    int dy[4] = {0, 1, 0, -1};
    
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int res = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] == 1) {
                    res = max(res, dfs(grid, i, j));
                }
            }
        }
        return res;
    }
    
    int dfs(vector<vector<int>>& grid, int x, int y) {
        int area = 1;
        grid[x][y] = 0;
        for (int i = 0; i < 4; i++) {
            int a = x + dx[i], b = y + dy[i];
            if (a >= 0 && a < grid.size() && b >= 0 && b < grid[0].size() && grid[a][b] == 1) {
                area += dfs(grid, a, b);
            }
        }
        return area;
    }
};
